artifacts/reader.py

```diff
@@ -227,6 +195,49 @@ def _ReadSupportedOS(self, definition_values, definition_object, name):
 
     definition_object.supported_os = supported_os
 
+  def _ReadSources(self, artifact_definition_values, artifact_definition, name):
+    """Reads the artifact definition sources.
+
+    Args:
+      artifact_definition_values: dictionary containing the artifact definition
+                                  values.
+      artifact_definition: the artifact definition (instance of
+                           ArtifactDefinition).09
+
+    Raises:
+      FormatError: if the type indicator is not set or unsupported,
+                   or if required attributes are missing.
+    """
+    sources = artifact_definition_values.get(u'sources')
+    if not sources:
+      raise errors.FormatError(
+          u'Invalid artifact definition: {0} missing sources.'.format(name))
+
+    for source in sources:
+      type_indicator = source.get(u'type', None)
+      if not type_indicator:
+        raise errors.FormatError(
+            u'Invalid artifact definition: {0} source type.'.format(name))
+
+      attributes = source.get(u'attributes', None)
+
+      try:
+        source_type = artifact_definition.AppendSource(type_indicator,
+                                                       attributes)
+      except errors.FormatError as exception:
+        raise errors.FormatError(
+            u'Invalid artifact definition: {0}. {1}'.format(name, exception))
+
+      # TODO: deprecate these left overs from the collector definition.
+      if source_type:
+        source_type.conditions = source.get(u'conditions', [])
+        source_type.returned_types = source.get(u'returned_types', [])
+        self._ReadSupportedOS(source, source_type, name)
+        if not set(source_type.supported_os) <= set(
```

Comment:
this condition is hard to reason about, I prefer the original code which has one more line but is much more readable.